libproxy/libproxyd.cpp

```diff
@@ -0,0 +1,177 @@
+/*******************************************************************************
+ * libproxy - A library for proxy configuration
+ * Copyright (C) 2015 Dominique Leuenberger <dimstar@opensuse.org>
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ ******************************************************************************/
+
+#include <gio/gio.h>
+#include <stdlib.h>
+#include <stdio.h>
+
+/* Import libproxy API */
+#include <proxy.h>
+
+/* ---------------------------------------------------------------------------------------------------- */
+
+static GDBusNodeInfo *introspection_data = NULL;
+
+/* Introspection data for the service we are exporting */
+static const gchar introspection_xml[] =
+  "<node>"
+  "  <interface name='org.libproxy.proxy'>"
+  "    <method name='query'>"
+  "      <annotation name='org.libproxy.proxy' value='OnMethod'/>"
+  "      <arg type='s' name='url'      direction='in'/>"
+  "      <arg type='as' name='response' direction='out'/>"
+  "    </method>"
+  "  </interface>"
+  "</node>";
+
+/* ---------------------------------------------------------------------------------------------------- */
+
+/* proxyFactory object is cached for the entire life-time of the dbus service */
+pxProxyFactory *pf;
+
+static void
+handle_method_call (GDBusConnection       *connection,
+                    const gchar           *sender,
+                    const gchar           *object_path,
+                    const gchar           *interface_name,
+                    const gchar           *method_name,
+                    GVariant              *parameters,
+                    GDBusMethodInvocation *invocation,
+                    gpointer               user_data)
+{
+
+  const gchar *url;
+  if (g_strcmp0 (method_name, "query") == 0)
+    {
+      g_variant_get (parameters, "(&s)", &url);
+      gchar **proxies = px_proxy_factory_get_proxies(pf, url);
+
+      GVariantBuilder *result = g_variant_builder_new (G_VARIANT_TYPE ("as"));
+
+      for (int j=0; proxies[j]; j++)
+      {
+          g_variant_builder_add(result, "&s", proxies[j]);
```

Comment:
I think the & here is ignored, because you are creating a GVariant, not getting data from it? (I know that with g_variant_get(), the & means you don't need to free the string.)